File: lua/cartpole_mdp.lua
local CartpoleMDP = torch.class('CartpoleMDP')

function CartpoleMDP:__init()
  self.py = require 'fb.python'
  self.py.exec([=[
from rllab.mdp.box2d.cartpole_mdp import CartpoleMDP
from rllab.mdp.normalized_mdp import normalize
mdp = normalize(CartpoleMDP())
  ]=])
end

function CartpoleMDP:reset()
  local ret = self.py.eval('mdp.reset()')
  return ret[1], ret[2]
end

function CartpoleMDP:step(state, action)
  local ret = self.py.eval('mdp.step(state, action)', {state = state, action = action})
  return ret[1], ret[2], ret[3], ret[4]
end

function CartpoleMDP:actionDim()
  return self.py.eval('mdp.action_dim')
end

function CartpoleMDP:observationShape()
  return self.py.eval('mdp.observation_shape')
end

function CartpoleMDP:actionBounds()
  local ret = self.py.eval('mdp.action_bounds')
  return ret[1], ret[2]
end

File: rllab/mdp/mujoco_1_22/cartpole_mdp.py
from rllab.mdp.mujoco_1_22.mujoco_mdp import MujocoMDP
import numpy as np
from rllab.core.serializable import Serializable
from rllab.misc import autoargs
from rllab.misc.overrides import overrides


class CartpoleMDP(MujocoMDP, Serializable):

    FILE = 'cartpole.xml'

    def __init__(
            self,
            *args, **kwargs):
        self.frame_skip = 1
        super(CartpoleMDP, self).__init__(*args, **kwargs)
        Serializable.__init__(
            self, *args, **kwargs)

    @overrides
    def get_current_obs(self):
        return np.concatenate([
            self.model.data.qpos[:1], # cart x pos
            np.sin(self.model.data.qpos[2:]), # link angles
            np.cos(self.model.data.qpos[2:]),
            np.clip(self.model.data.qvel, -10, 10),
            np.clip(self.model.data.qfrc_constraint, -10, 10)]
        ).reshape(-1)

    @overrides
    def step(self, state, action):
        next_state = self.forward_dynamics(state, action, restore=False)
        next_obs = self.get_obs(next_state)
        x, _, y = self.model.data.site_xpos[0]
        r = -(0.01*x**2 + (y-2)**2)
        return next_state, next_obs, r, False

    @overrides
    def reset(self):
        self.model.data.qpos = self.init_qpos
        self.model.data.qvel = self.init_qvel
        self.model.data.qacc = self.init_qacc
        self.model.data.ctrl = self.init_ctrl
        self.model.forward()
        self.current_com = self.model.data.com_subtree[0]
        self.dcom = np.zeros_like(self.current_com)
        self.current_state = self.get_current_state()
        return self.get_current_state(), self.get_current_obs()
